### PR TITLE
Add reshape int8 op

### PR DESCRIPTION
**Pull request by wozna**
This PR enables quantization of reshape2 operator. Type int8 is not added in MLKDNN operator implementation, but in global implementation.  
Reshape and transpose ops are quantized only if previous operator or next operator is quantized. Thanks that we avoid situation where there is one int8 operators between two fp32 operators.

### CODE REVIEW

File: paddle/fluid/inference/api/mkldnn_quantizer_config.cc [hunk 1]
```diff
@@ -42,6 +42,12 @@ MkldnnQuantizerConfig::MkldnnQuantizerConfig() {
   rules_["fc"]["W"] = ScaleAlgo::MAX_CH_T;
   rules_["fc"]["Bias"] = ScaleAlgo::NONE;
   rules_["fc"]["Out"] = ScaleAlgo::KL;
+
+  rules_["reshape2"]["X"] = ScaleAlgo::KL;
+  rules_["reshape2"]["Shape"] = ScaleAlgo::NONE;
+  rules_["reshape2"]["ShapeTensor"] = ScaleAlgo::NONE;
+  rules_["reshape2"]["XShape"] = ScaleAlgo::NONE;
+  rules_["reshape2"]["Out"] = ScaleAlgo::NONE;
```
Comment (wojtuss) [inline paddle/fluid/inference/api/mkldnn_quantizer_config.cc:55]: A comment explaining why we do not need to calculate the scale for `reshape2`'s output could be helpful here. Could be added also to `transpose2`.

File: paddle/fluid/operators/reshape_op.cc [hunk 2]
```diff
@@ -245,6 +245,13 @@ class ReshapeOpMaker : public framework::OpProtoAndCheckerMaker {
         "It has the lowest priority compare with Input(Shape) and "
         " Input(ShapeTensor).")
         .SetDefault({});
+    /* int8 parameters */
+    AddAttr<bool>("use_quantizer",
+                  "(bool, default false) "
+                  "Set to true for operators that should be quantized and use "
+                  "int8 kernel. "
+                  "Only used on CPU.")
```
Comment (wojtuss) [inline paddle/fluid/operators/reshape_op.cc]: I think "Used only on CPU." is more grammatically correct.

File: paddle/fluid/operators/reshape_op.cc [hunk 3]
```diff
@@ -570,15 +580,20 @@ REGISTER_OPERATOR(reshape2_grad_grad, ops::Reshape2DoubleGradOp,
...
 REGISTER_OP_CPU_KERNEL_FUNCTOR(reshape2, float, ops::ReshapeKernel, double,
-                               ops::ReshapeKernel, int, ops::ReshapeKernel,
-                               int64_t, ops::ReshapeKernel);
+                               ops::ReshapeKernel, int8_t, ops::ReshapeKernel,
+                               uint8_t, ops::ReshapeKernel, int,
+                               ops::ReshapeKernel, int64_t, ops::ReshapeKernel);
 REGISTER_OP_CPU_KERNEL_FUNCTOR(reshape2_grad, float, ops::ReshapeGradKernel,
                                double, ops::ReshapeGradKernel, int,
+                               ops::ReshapeGradKernel, int8_t,
+                               ops::ReshapeGradKernel, uint8_t,
```
Comment (wojtuss) [inline paddle/fluid/operators/reshape_op.cc]: We do not provide gradient kernels for the `reshape(2)` ops.

File: paddle/fluid/operators/reshape_op.cc [hunk 4]
```diff
@@ -553,9 +560,12 @@ REGISTER_OPERATOR(reshape_grad, ops::ReshapeGradOp,
 
 REGISTER_OP_CPU_KERNEL_FUNCTOR(reshape, float, ops::ReshapeKernel, double,
                                ops::ReshapeKernel, int, ops::ReshapeKernel,
-                               int64_t, ops::ReshapeKernel);
+                               int8_t, ops::ReshapeKernel, uint8_t,
```
Comment (wojtuss) [inline paddle/fluid/operators/reshape_op.cc]: I think we do not need the int8 kernels for the `reshape` op. We are interested in quantizing `reshape2` only now.

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass_tester.cc [hunk 5]
```diff
@@ -370,6 +370,122 @@ TEST(CpuQuantizePass, transpose) {
...
+        dequantize_nodes_count++;
+      }
+    }
+  }
+  EXPECT_EQ(transpose_nodes_count, transpose_count);
+  EXPECT_EQ(reshape_nodes_count, reshape_count);
+  EXPECT_EQ(conv_nodes_count, conv_count);
+  EXPECT_EQ(quantize_nodes_count, quant_count);
+  EXPECT_EQ(dequantize_nodes_count, dequant_count);
+  EXPECT_EQ(original_nodes_num + added_nodes_count, current_nodes_num);
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass_tester.cc:447]: Please, verify also that the scales used for quantization of `reshape2`'s input and output are the same.

File: paddle/fluid/inference/api/mkldnn_quantizer_config.cc (hunk 1 above)
Comment (wozna) [reply paddle/fluid/inference/api/mkldnn_quantizer_config.cc:55]: Done

File: paddle/fluid/operators/reshape_op.cc (hunk 2 above)
Comment (wozna) [reply paddle/fluid/operators/reshape_op.cc]: Done

File: paddle/fluid/operators/reshape_op.cc (hunk 3 above)
Comment (wozna) [reply paddle/fluid/operators/reshape_op.cc]: You're right. Done

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass_tester.cc (hunk 5 above)
Comment (wozna) [reply paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass_tester.cc:447]: Done

File: paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass_tester.cc [hunk 6]
```diff
@@ -365,11 +364,120 @@ TEST(CpuQuantizePass, transpose) {
...
+        std::cout << op_name << " \n";
+        if (op_name != "Dequantize1") {
+          dequant_scale = boost::get<float>(op->GetAttr("Scale"));
+          EXPECT_EQ(dequant_scale, scale)
+              << "Scale for node '" + op->Type() + "'.";
+        }
+      }
+    }
+  }
+  EXPECT_EQ(dequant_scale, quant_scale);
```
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass_tester.cc]: The equality of those scales is already verified in the above `quantize`'s and `dequantize`'s scales check.

Comment (wozna) [reply paddle/fluid/framework/ir/mkldnn/cpu_quantize_pass_tester.cc]: You are right. It is corrected.

**Pull request merged**